WordPress/gutenberg
# Fix undo trap in Gallery block when transforming from shortcode

## What?

Related to https://github.com/WordPress/gutenberg/issues/8119.

Fix "undo trap" in the Gallery block when transforming from shortcode. 

## Why?


See the original issue for more info. When pasting shortcode like `[gallery ids="1,2,3"]` to the editor, the shortcode will be transformed into a gallery block, but it's unable to undo it by using the Undo/Redo button.

## How?

The key here is that the `updateImages( shortCodeImages )` call in the `useEffect` will schedule an update to set the attributes for its children and thus creating new undo levels automatically after the first render. The `useEffect` here (along with other `useEffect`s in the same component) is an [anti-pattern](https://beta.reactjs.org/learn/you-might-not-need-an-effect) and is redundant.

The solution is to get rid of the `useEffect` and directly inject the inner `core/image` blocks instead. It requires us to update the shortcode converter to include a `__experimentalTransform` function so that we can define the transformed `innerBlocks`. Similar patterns can be seen in other transformation methods like the blocks transformer. I think it makes sense to add it to the shortcode transformer too.

## Testing Instructions




1. Add some photos to the Media Library and obtain their `id`s.
2. Construct a paragraph with shortcode like below (replace the `ids` with yours):
    ```
    [gallery ids="12"]
    Something
    [gallery ids="13,14"]
    ```
3. Copy the content and paste it into the editor.
4. See that the shortcodes get transformed into Gallery blocks.
5. Click the "Undo" button in the toolbar, the content should be emptied.
6. Click the "Redo" button in the toolbar, the gallery blocks should be restored.

## Screenshots or screencast 

Before | After
--- | ---
<video src="https://user-images.githubusercontent.com/7753001/176141217-bd9ac5db-2a95-45b8-a379-97d7bf412ef9.mp4"></video> | <video src="https://user-images.githubusercontent.com/7753001/176141280-82c968f0-17b6-4c8e-a459-92b34a22e1a1.mp4"></video>
